pkg/agent/nodeportlocal/k8s/annotations.go

```diff
@@ -45,106 +46,79 @@ func toJSON(serialize interface{}) string {
+		// no updates required to the Pod
+		return false
 	}
+	return true
+}
 
+func removeFromNPLAnnotation(annotations []NPLAnnotation, containerPort int) []NPLAnnotation {
 	for i, ann := range annotations {
 		if ann.PodPort == containerPort {
 			annotations = append(annotations[:i], annotations[i+1:]...)
 			break
 		}
 	}
+	return annotations
+}
 
-	current[NPLAnnotationKey] = toJSON(annotations)
-	pod.Annotations = current
+func (c *NPLController) updatePodNPLAnnotation(pod *corev1.Pod, annotations []NPLAnnotation) error {
+	if err := patchPod(annotations, *pod, c.kubeClient); err != nil {
+		klog.Warningf("Unable to patch NPL annotations for Pod %s/%s: %s", pod.Namespace, pod.Name, err.Error())
+	}
+	klog.V(2).Infof("Successfully updated annotation for Pod %s/%s", pod.Namespace, pod.Name)
+	return nil
 }
 
-// RemoveNPLAnnotationFromPods removes npl annotations from all Pods.
-func (c *NPLController) RemoveNPLAnnotationFromPods() {
-	nodeName, err := env.GetNodeName()
-	if err != nil {
-		klog.Warningf("Failed to get Node's name, NodePortLocal annotation cannot be removed for Pods scheduled to this Node")
-		return
+func patchPod(value []NPLAnnotation, pod corev1.Pod, kubeClient clientset.Interface) error {
+	payloadValue := make(map[string]*string)
+	if len(value) > 0 {
+		valueStr := string(toJSON(value))
+		payloadValue[NPLAnnotationKey] = &valueStr
+	} else {
+		payloadValue[NPLAnnotationKey] = nil
 	}
-	podList, err := c.kubeClient.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{
-		FieldSelector:   "spec.nodeName=" + nodeName,
-		ResourceVersion: "0",
-	})
-	if err != nil {
-		klog.Warningf("Unable to list Pods, err: %v", err)
-		return
+
+	newPayload := map[string]interface{}{
+		"metadata": map[string]map[string]*string{
+			"annotations": payloadValue,
+		},
 	}
-	for i, pod := range podList.Items {
-		if _, exists := pod.Annotations[NPLAnnotationKey]; !exists {
-			continue
-		}
-		removePodAnnotation(&podList.Items[i])
-		c.updatePodAnnotation(&podList.Items[i])
+
+	payloadBytes, _ := json.Marshal(newPayload)
+	if _, err := kubeClient.CoreV1().Pods(pod.Namespace).Patch(context.TODO(), pod.Name, types.MergePatchType,
+		payloadBytes, metav1.PatchOptions{}); err != nil {
+		return fmt.Errorf("unable to update Annotation for Pod %s/%s, %s", pod.Namespace,
+			pod.Name, err.Error())
 	}
-	klog.Infof("Removed all NodePortLocal annotations from all Pods")
+	return nil
 }
 
-func (c *NPLController) updatePodAnnotation(pod *corev1.Pod) error {
-	if _, err := c.kubeClient.CoreV1().Pods(pod.Namespace).Update(context.TODO(), pod, metav1.UpdateOptions{}); err != nil {
-		klog.Warningf("Unable to update annotation for Pod %s/%s, error: %v", pod.Namespace, pod.Name, err)
-		return err
+// CleanupNPLAnnotationForPod removes the NPL Annotation from the Pod's Annotations map entirely.
+func CleanupNPLAnnotationForPod(kubeClient clientset.Interface, pod corev1.Pod) error {
```

Comment:
```suggestion
func CleanupNPLAnnotationForPod(kubeClient clientset.Interface, pod *corev1.Pod) error {
```
to avoid a copy?

Reply:
Sure, the idea was to not use a pointer because we weren't changing anything for that pod. But yeah, we should avoid a copy too. Changed it.